nucypher/keystore/keystore.py

```diff
@@ -160,30 +159,52 @@ def attach_kfrag_to_saved_arrangement(self, alice, id_as_hex, kfrag, session=Non
         policy_arrangement.kfrag = bytes(kfrag)
         session.commit()
 
-    def add_workorder(self, bob_verifying_key, bob_signature, arrangement_id, session=None) -> Workorder:
+    def save_workorder(self, bob_verifying_key, bob_signature, arrangement_id, session=None) -> Workorder:
         """
         Adds a Workorder to the keystore.
         """
         session = session or self._session_on_init_thread
-        bob_verifying_key = self.add_key(bob_verifying_key)
-        new_workorder = Workorder(bob_verifying_key.id, bob_signature, arrangement_id)
+
+        # Get or Create Bob Verifying Key
+        fingerprint = fingerprint_from_key(bob_verifying_key)
+        key = session.query(Key).filter_by(fingerprint=fingerprint).first()
+        if not key:
+            key = self.add_key(key=bob_verifying_key)
+
+        new_workorder = Workorder(bob_verifying_key_id=key.id,
+                                  bob_signature=bob_signature,
+                                  arrangement_id=arrangement_id)
 
         session.add(new_workorder)
         session.commit()
-
         return new_workorder
 
-    def get_workorders(self, arrangement_id: bytes, session=None) -> Workorder:
+    def get_workorders(self,
+                       arrangement_id: bytes = None,
+                       bob_verifying_key: bytes = None,
+                       session=None
+                       ) -> List[Workorder]:
         """
         Returns a list of Workorders by HRAC.
         """
         session = session or self._session_on_init_thread
-
-        workorders = session.query(Workorder).filter_by(arrangement_id=arrangement_id)
-
-        if not workorders:
-            raise NotFound("No Workorders with {} HRAC found.".format(arrangement_id))
-        return workorders
+        query = session.query(Workorder)
+
+        if not arrangement_id and not bob_verifying_key:
+            workorders = list(query.all())
+        else:
+            if arrangement_id:
+                workorders = query.filter_by(arrangement_id=arrangement_id)
+            elif bob_verifying_key:
+                fingerprint = fingerprint_from_key(bob_verifying_key)
+                bvk = session.query(Key).filter_by(fingerprint=fingerprint).first()
+                workorders = query.filter_by(bob_verifying_key_id=bvk.id)
+            else:
+                raise ValueError
```

Comment:
Does the code ever get to this `else` block to raise the `ValueError`? i.e. Is this necessary?

Reply:
I think you can replace `elif bob_verifying_key:` with `else:` and then you should be able to get rid of the `else: raise ValueError` block. Hopefully that shuts up your linter hehe 😛

Reply:
:heavy_check_mark: